(#1754) Update documentation about contributing to repo (#2155)
chocolatey-community/chocolatey-packages: CONTRIBUTING.md

# Contributing Guidelines

Thank you for your interest in contributing to the Chocolatey Community Packages repository.
This document details the expectations of what is needed when contributing to this repository, and what all contributors need to follow.

This repository presents **the latest and highest package standards**. The purpose of this repository is to provide packages that are:

- **Stable** - When possible, current and earlier versions should generally work so you could depend on them. The most stable packages are those that are embedded.
- **High quality** - Packages should be resilient and should provide parameters where adequate.
- **Free** - Packages should be generally usable by anybody without any prerequisites.

> To aachieve these goals we are using the following priorities when adding new or maintaining existing packages:

Reviewer: ```suggestion
To achieve these goals we are using the following priorities when adding new or maintaining existing packages:
```

1. Cross platform FOSS packages.
2. Windows only FOSS packages.
3. Freeware packages.
4. Commercial packages with unrestricted trials.

The following rules also apply:

> 1. Whe npackages have the same priorities, software with higher number of users will generally be considered more important.

Reviewer: ```suggestion
1. When packages have the same priorities, software with higher number of users will generally be considered more important.
```

  - [Chocolatey Scripts](#chocolatey-scripts)
    - [UI Automation](#ui-automation)
      - [Work on all locales](#work-on-all-locales)
      - [Avoid brittle scripts](#avoid-brittle-scripts)
    - [Source Files](#source-files)
      - [Encoding](#encoding)
      - [Code style](#code-style)
  - [Update Script](#update-script)
    - [Use `UseBasicParsing`](#use-usebasicparsing)
    - [Do not download large files](#do-not-download-large-files)
- [FAQ](#faq)
  - [Why do we automatically close issues?](#why-do-we-automatically-close-issues)
  - [Why do not the repository maintainers fix the issues?](#why-do-not-the-repository-maintainers-fix-the-issues)
  - [How can I create a fix version of a package?](#how-can-i-create-a-fix-version-of-a-package)

</details>
<!-- END doctoc generated TOC please keep comment here to allow auto update -->

<!-- markdownlint-enable -->

## Etiquete

Make sure that all communications you make on this repository follows our [CODE OF CONDUCT](https://github.com/chocolatey-community/.github/blob/main/CODE_OF_CONDUCT.md). Failure to adhere to this document may lead you to be banned from the repository.

> In addition to the code of conduct document, make sure that you do not make unneeded reference to other users when opening issues, discussions or pull requests, doing so may lead to the issue or pull request being ignored by these users instead.

Reviewer: It would be good to explicitly state that references means @-ing a user.

Reviewer 2: Its proper term is to  [mention](https://github.blog/2011-03-23-mention-somebody-they-re-notified/).

### Requesting enhancements or new functionality

Requesting new functionality to any package is not accepted as direct pull requets, nor as normal issues.
Any enhancement or functionality requests must be opened as a discussion.
A new feature or enhancement once accepted will be created as a issue with a reference to the discussion once it has been approved.
A feature  or enhancement will not be approved until it is known who will be working on the implementation, so if you wish to work on it make sure to mention this in the initial discussion body.

Issues created as a bug or as an outdated report may be closed immediately if it is determined to be a new functionality, or an ehancement to existing packages, you may then have to open a discussion about the package instead (there are times a maintainer may convert the issue to a discussion, but this is no guarantee).

### Issue closing

We have an automated bot that goes through the issues created on the repository that will mark issues as stale, and close any issues that has been marked as stale.

Most issues are considered when marking these as stale, with the exception of issues marked with one of the following labels:

> - `Security / CVE` - This one is not considered as we always want security related issues resolved.

Reviewer: ```suggestion
- `Security / CVE` - This label is ignored by the bot as we always want security related issues resolved.
```

- `0 - Backlog` - This one is not considered as it will be worked on by a repository maintainer, butit is not certain when it will be worked on.
- `1. - Read for work` - This one is not considered as it will be worked on by a repository maintainer, and a date when the work is to be started has been set.
- `2. - Working` - This one is not considered as a repository maintainer has already started his/her work for the issue.
- `3. - Review` - This one is not considered as the work to complete the issue has already been submitted as a pull request.

> - `5. - Push required` - This one is not considered as a package with the work for fixing the issue has not been submitted to community.chocolatey.org.

Author: If we have something marked as done, and it is still open, we have a bigger problem, so no shouldn't need to have it (I know you were joking, but still responding).

> - `5. - Push required` - This one is not considered as a package with the work for fixing the issue has not been submitted to community.chocolatey.org.

Reviewer 3: Fair!

When opening a new pull request it is required that you input all of the information requested by the pull request template being used.
When the pull request opened is for new features, enhancements or migrating a package it is required that a issue exists already before opening the pull request.
If no issue exists, or if an issue was opened after the pull request was opened it may be closed immediately without response.

Pull requests fixing a bug or an outdated package do not need to have a issue associated with it.

Do not open a pull request whene there is already an open pull request for the same issue or bug, as this will lead in your pull request becoming closed.

Your changes should also only affect a single package, except if the package in question also consists of a meta, `.install` and `.portable` packages. In this case all three of the packages can be submitted in the same pull request.
There may be other exceptions as well, but these will be on a case by case basis, and only when approved by a repository maintainer.

Make sure you read through the [Code Conventions](#code-conventions) as well before opening the pull request.

### Pull request title

> The pull request title should always be prefixed with the `(packageName)` where the package nam is replaced with the name of the this pull request is for.

Reviewer: ```suggestion
The pull request title should always be prefixed with the `(packageName)` where the package name is replaced with the name of the package this pull request is for.
```

**NOTE: Do not add a comment just to circumvent the automatic closure, this may lead the pull request to be closed immediately instead**

## Conventions / Guidelines

All packages in this repository are expected to pass all Requirement, Guideline and Suggestions checks on the Chocolatey Community Repository when it is possible.
Additionally, all packages are also expected to following the additional Conventions / Guidelines we have outlined in this section.

### Manual or automatic

When working on packages in this repository, a decision must be made whether the package should be an automatic or manual package.
In most cases it is preferable that a package is made automatic, as these can then be updated without doing manual changes when there is a new release of the
software that is being packaged.
The only time an automatic package should be considered to be a manual package is if the underlying software that is being packaged has not been updated for 3 or more years.
New packages contributed/migrated to this repository should always be created as automatic packages.

> All automatic packages are expected to make use of the [AU](https://github.com/chocolatey-community/chocolatey-au) module, and work in Windows PowerShell 5.x, PowerShell Core is not supported.

Reviewer 3: ```suggestion
All automatic packages are expected to make use of the [AU](https://github.com/chocolatey-community/chocolatey-au) module, and work in Windows PowerShell 5.x, as the builds do not run PowerShell Core.
```

Author: Here is the thing, it may run on PowerShell Core, but it may behave differently than we expect. So, would this change still be correct in this case?

Reviewer 3: I guess I just wanted to highlight the _reason_ for us not allowing scripts that _may_ work in Core, so that people would understand - AppVeyor runs 5, currently, and that's what actually produces and pushes our packages. What we actually care about is that it works correctly in 5, and on our CI.

```suggestion
All automatic packages should make use of the [AU](https://github.com/chocolatey-community/chocolatey-au) module, and must work in Windows PowerShell 5.x.
```

> The following metadata, and scripts conventions are for automatic packages, but manual packages are expected to follow the same rules when possible.

Reviewer 3: ```suggestion
The following metadata and script conventions are for automatic packages, but manual packages are expected to follow the same rules where possible.
```

### Package types

### Embedded packages

> _Embedded_ packages include the packaged software directly in the nupkg archive instead of downloading it. Only tools that allow redistribution in their license can be embedded and such packages must include two additional files in the directory `legal` - `VERIFICATION.txt` and `License.txt`.

Reviewer 3: ```suggestion
Embedded packages include the packaged software directly in the nupkg archive instead of downloading it. Only tools that allow redistribution in their license can be embedded and such packages must include two additional files in the `legal` directory within the source folder and shipped package - `VERIFICATION.txt` and `LICENSE.txt`.
```

Author: I think keeping the italic is still valid here, especially since we aren't removing it from the other sections (Sem-Embedded, Remote, Etc)

```suggestion
_Embedded_ packages include the packaged software directly in the nupkg archive instead of downloading it. Only tools that allow redistribution in their license can be embedded and such packages must include two additional files in the `legal` directory within the source folder and shipped package - `VERIFICATION.txt` and `LICENSE.txt`.
```

Reviewer 3: Sure! Possibly at the time I hadn't got down to those lines yet, and thought it was an unusual highlight. Makes more sense with that context!

> Its **recommended to create embedded packages** because they don't depend on vendor site working and substantially reduce the network related problems - 404 (file not found) problem and potential vendor bandwidth leaching issues are completely solved by embedding.

Reviewer 3: ```suggestion
It is **recommended to create embedded packages** because they don't depend on an external site working, and substantially reduce the potential for network related problems - 404 (file not found) problems, and potential vendor bandwidth leaching issues are completely solved by embedding the binaries within the package.
```

Binary files can not be generally checked in into this repository because that will bloat it too much.

> The repository is ignoring binary files via `.gitignore`. Automatic packages use AU functions to produce correct package that includes binaries during automatic update procedure.

Reviewer 3: ```suggestion
The repository has a `.gitignore` which excludes many popular binaries. Automatic packages use AU functions to produce packages that include binaries during the automatic update procedure.
```

See the following packages as an example: [qbittorent](https://github.com/chocolatey-community/chocolatey-packages/tree/master/automatic/qbittorrent),
[7zip.install](https://github.com/chocolatey-community/chocolatey-packages/tree/master/automatic/7zip.install), [transifex-client](https://github.com/chocolatey-community/chocolatey-packages/tree/master/automatic/transifex-client).

For software that explicitly doesn't allow redistribution via adequate license then one may **contact the vendor**, ask for the redistribution rights and provide proof in the package in the form of:

- PDF of a signed license
- signed letter
- PDF of an email chain granting that permission

> As an example take a look at the [activepresenter](https://github.com/chocolatey-community/chocolatey-packages/tree/master/automatic/activepresenter/legal) package. Embeding non-allowed binaries may have [legal repercussions](https://docs.chocolatey.org/en-us/information/legal).

Reviewer 3: ```suggestion
As an example take a look at the [activepresenter](https://github.com/chocolatey-community/chocolatey-packages/tree/master/automatic/activepresenter/legal) package. Embedding non-allowed binaries may have [legal repercussions](https://docs.chocolatey.org/en-us/information/legal).
```

> **NOTE**: 200MB is the maximum size of the package. Larger tools must be downloaded from a vendor site or mirror.

Reviewer 3: ```suggestion
**NOTE**: 200MB is the maximum size of a package on the Chocolatey Community Repository. Larger tools must be downloaded from a vendor site or mirror.
```

### Semi-Embedded packages

> _Semi-Embedded_ packages are packages that only partially embeds the software inside of the nupkg archive.

Reviewer 3: ```suggestion
_Semi-Embedded_ packages are packages that only partially embed the software inside of the nupkg archive.
```

> This is a common approach being used when the size of the package has become to large, and it includes both 32bit and 64bit softwares.

Reviewer 3: ```suggestion
This is a common approach being used when the size of the package has become too large, and the package currently includes both 32bit and 64bit versions of the software.
```

> In these cases the best approach when possible is to keep the 64bit softwares still embedded, while downloading any 32bit softwares during installation if needed.

Reviewer 3: ```suggestion
In these cases the best approach when possible is to keep the 64bit software still embedded, while downloading any 32bit software during installation if needed (as Windows is increasingly popular in 64bit arch).
```

### Remote packages

> _Remote_ packages does not include any packaged software directly in the nupkg archive, instead it downloads what is necessary when a user installs the package.

Reviewer 3: ```suggestion
_Remote_ packages do not include any packaged software directly in the nupkg archive, instead it downloads what is necessary when a user installs the package.
```

> This is the most common type seen on the Chocolatey Community Repository, and is the only valid approach if the software being packaged does not allow for it to be redistributed and the software authors are not granting and redistribution rights.

Reviewer 3: ```suggestion
This is the most common type seen on the Chocolatey Community Repository, and is the only valid approach if the software being packaged does not allow for it to be redistributed and the software authors are not willing to grant redistribution rights.
```

Additionally, when a package becomes too large to be uploaded to Chocolatey Community Repository it is valid to change an embedded package to a remote package.

### Metadata

> This section details the conventions and guideliens that are related to changing or updating metadata for each package.

Reviewer 3: ```suggestion
This section details the conventions and guidelines that are related to changing or updating metadata for each package.
```

> Metadata are explicitly only the information that is defined for the package in its related `.nuspec` file.

Reviewer 3: ```suggestion
Metadata is explicitly only the information that is defined for the package in its related `.nuspec` file.
```

#### Naming

> For automatic packages the name of the package is taken from its root directory name, if the original idenifier of the existing package is not in lowercase this needs to be overridden in an `update.ps1` script.

Reviewer 3: ```suggestion
For automatic packages the name of the package is taken from its root directory name. If the original identifier of the existing package is not fully lowercase this needs to be overridden in an `update.ps1` script.
```

For manual packages this is defined directly in the package nuspec file, in general this should also be in lowercase but for existing packages it should use the same casing as the original identifier.

> This root directory name should always be in lower case and follow the official Chocolatey naming conventions: https://docs.chocolatey.org/en-us/create/create-packages#naming-your-package.

Reviewer 3: ```suggestion
This root directory name should always be in lower case and follow the official [Chocolatey naming conventions](https://docs.chocolatey.org/en-us/create/create-packages#naming-your-package).
```

#### Dependency versions

> 1. When taking a dependency on an extension, a minimum version of that dependency must be specified. Without this, any version satisfies the dependency. That means unless someone upgrades the extension outside of their process or incidentally install some package that uses newer version explictly set, it will not automatically upgrade to the latest version. Changes to a package that does not specify a minimum version, or exact version is will not be accepted.

Reviewer 3: ```suggestion
1. When taking a dependency on an extension, a minimum version of that dependency must be specified. Without this, any version satisfies the dependency. That means unless someone upgrades the extension outside of their process or incidentally install some package that uses an explicitly set newer version, it will not automatically upgrade to the latest version. Changes to a package that do not specify a minimum version, or exact version, not be accepted.
```

> 2. When creating a dependency for virtual packages, a exact version range should be used for the dependent package (_.install_ or _.portable), which should be the same as that of the virtual package. An exact version range in the metadata file will look like `[1.0.0, 1.0.0]`.

Reviewer 3: ```suggestion
2. When creating a dependency for virtual packages, an exact version range should be used for the dependent package (_.install_ or _.portable). This version should be the same as that of the virtual package. An exact version in the metadata file will look like `[1.0.0]`.
```

#### Conform to guidelines

Conform with the [official package creation guidelines](1754-Update-Guidelines-and-code-to-reflect-new-procedures) and take a look at the [quick start guide](https://docs.chocolatey.org/en-us/create/create-packages-quick-start) on how to create packages.

You should also know how to [deprecate a package](https://docs.chocolatey.org/en-us/create/create-packages-quick-start)

#### Obligatory metadata

> When possible, all elements in the nuspec file is obligated to be made use of. This includes elements like `packageSourceUrl`, `projectSourceUrl`, `docsUrl`, `bugTrackerUrl`, `releaseNotes`, `licenseUrl` and `iconUrl`, even when one or more of these are made optional on Chocolatey Community Repository itself. If the software itself does not have any sufficient URLs to be used for these examples, then it is fine to be omitted.

Reviewer 3: ```suggestion
When possible, all possible elements in the nuspec file should be filled. This includes elements like `packageSourceUrl`, `projectSourceUrl`, `docsUrl`, `bugTrackerUrl`, `releaseNotes`, `licenseUrl` and `iconUrl` - even when one or more of these are made optional on Chocolatey Community Repository itself. If the software itself does not have any sufficient URLs to be used for these examples then it can be omitted.
```

#### Obligatory tags and categories

All tags and categories should be in lowercase, and should be in the following order.
The tags specified should always be space delimited.

- The first specified tag should always be a dash delimited name of the identifier of the package (excluding the suffixes `.install`, `.portable`, `.commandline` and `.app`).
- The second specified tag should be the category that the packaged software falls under. Only the following list of tags are allowed to be used as the category tag (_NOTE: Existing packages may be missing a category tag, if you change a package that does not make use of one, you will be asked to include a category tag_)

>   - `addon` (For packaged softwares that are plugins or addons for other applications)

Reviewer 3: ```suggestion
  - `addon` (for packaged software that is a plugin or addon for another application)
```

>   - `browser` (Specifically for browser implementations, like Google Chrome, Firefox, Microsoft Edge, Brave, Etc)

Reviewer 3: ```suggestion
  - `browser` (Specifically for browser implementations like Google Chrome, Firefox, Microsoft Edge, Brave, etc.)
```

  - `client`
  - `driver`
  - `editor`
  - `extension` (only for Chocolatey Extensions)
  - `games`
  - `productivity`
  - `programming`
  - `server`
  - `utility`
  - `web`
- The third tag denotes what type of application is being packages when it comes to license related information. These tags can be summarized with the following:
  - `foss` (The packaged application is a free application with its source freely viewable by anyone)
  - `freeware` (The packaged application is a free application, but its source is not available to be viewed)

>   - `trial` (The packaged application is a commercial or payed application, but includes a limited free trial that can be used)

Reviewer 3: ```suggestion
  - `trial` (The packaged application is a commercial or paid application, but includes a limited free trial that can be used)
```

- If the application is available on multiple platforms, the fourth tag should be `cross-platform`.
- More tags can be added and is recommended to be added when possible, these tags should not be of any of the previously mentioned tags used for categorization or licensing.

#### Description

The description of the package can usually be acquired from the software authors website, and should at a minimum include enough information about the application for those that knows nothing of them to be able to give a decision if the application is for them or not.

Additionally there are a few descriptive headers that are required to be specified in the description as well.

| Header Name          | Meaning |
|----------------------|---------|
| `Features`           | Bullet list that sumarrizes the availbale functionality of the packaged application |

> | `Package Parameters` | Bullet list of the available package parameters that can be used when installing the package, this section can be omitted if there is no package parameters |

Reviewer 3: ```suggestion
| `Package Parameters` | Bullet list of the available package parameters that can be used when installing the package. This section can be omitted if there are no package parameters |
```

> | `Notes`              | Bullet list with any special information to the user about the software and/or package that they need to know about |

Reviewer 3: ```suggestion
| `Notes`              | Bullet list with any special information about the software and/or package that the user should know about, e.g. recent breaking changes to the install, or unusual edge cases |
```

#### Maintainers

> Maintainers of the package in our case are users that have edited a package, either by being the main maintainer of the package or contributed one or more changes to the package.

Reviewer 3: ```suggestion
Maintainers of the package, in our case, are folk that have edited a package - either by being the original maintainer of the package, or by contributing one or more changes to the package after it has been added to the repository.
```

> The metadata file should always include both current and historical maintainers of the package, if you are editing a package make sure to add yourself as the last maintainer colon delimeted list.

Reviewer 3: ```suggestion
The metadata file should always include both current and historical maintainers of the package, if you are editing a package make sure to add yourself, using a comma-delimited list format.
```

If the user `chocolatey-community` is not already specified as the maintainer of the package, this user must be added as the first maintainer in the `owners` element.
One or more maintainers of the maintainers listed in the `owners` element should be equal to a similar name listed in our `CODE_OWNERS` file, and will be considered the main maintainer of the package.

#### Icons

A package must have an icon available if the packaged software has an icon. This icon must be named the same as the package and is placed in the [icons](https://github.com/chocolatey-community/chocolatey-packages/tree/master/icons) directory.

> If the package name ends with either `.install`, `.portable` those suffixes may be ignored in the icon name.

Reviewer 3: ```suggestion
If the package name ends with either `.install` or `.portable`, the suffix may be ignored in the icon name.
```

> When an icon is added to this folder with the correct name, it wwill **automatically** be set in the metadata file and the Readme file when our build server updates the package if the metadata file contains an `<IconUrl>` element.

Reviewer 3: ```suggestion
When an icon is added to this folder with the correct name, it will **automatically** be set in the metadata file and the Readme file when our build server updates the package, if the metadata file contains an `<IconUrl>` element.
```

> **IMPORTANT: If there is no icon available, the comment `<!-- IconUrl: Skip check -->` is required to be specified in the metadata file**

Reviewer 3: ```suggestion
**IMPORTANT: If there is no icon available, the comment `<!-- IconUrl: Skip check -->` should be added to the metadata file**
```

#### Chocolatey compatibility

All packages are expected to be compatible with the version of Chocolatey CLI that was available 12 months earlier than the last change of the package.
With that, it means that if you make a change to the nuspec, or one of the package scripts on January 12th, 2023 it should be compatible with Chocolatey CLI 0.11.3 and later released versions.
If that is not possible, and the package needs a version of Chocolatey CLI that was released later, it will need to include a dependency for that Chocolatey CLI version.

### Chocolatey Scripts

> All PowerShell scripts are expected to be compatible with PowerShell v2 and later (not PowerShel Core). If this is not possible, a dependency on the package `PowerShell` is required with its minimum version set to the earliest version that the scripts in the package is expected to work with.

Reviewer 3: ```suggestion
All PowerShell scripts are expected to be compatible with PowerShell v2 and later (not PowerShell Core). If this is not possible, a dependency on the package `PowerShell` is required with its minimum version set to the earliest version that the scripts in the package is expected to work with.
```

> **TODO: This section needs to be updated, and is a copy-paste from earlier guidelines**

Reviewer 3: Just flagging this, either for removal (if the below is alright) or action.

Author: Yeah, I think it can be removed

```suggestion
```

#### UI Automation

Some installers do not provide silent arguments and can be difficult to automate.

> If the package needs to perform tasks that cannot be done with command line switches, e.g. clicking away a prompt during installation that cannot be suppressed like for example in the [dropbox](https://chocolatey.org/packages/dropbox) package, you can use [AutoHotkey](https://chocolatey.org/packages/autohotkey.portable). Make [autohotkey.portable](https://chocolatey.org/packages/autohotkey.portable) a dependency of the package. Prefer AutoHotkey over [AutoIt](https://chocolatey.org/packages/autoit.commandline), because AutoHotkey is more lightweight, FOS and more actively developed than AutoIt.

Reviewer 3: ```suggestion
If the package needs to perform tasks that cannot be done with command line switches, e.g. clicking away a prompt during installation that cannot be suppressed as in the [dropbox](https://chocolatey.org/packages/dropbox) package, you can use [AutoHotkey](https://chocolatey.org/packages/autohotkey.portable). Make [autohotkey.portable](https://chocolatey.org/packages/autohotkey.portable) a dependency of the package. In general, the community maintainers prefer AutoHotkey over [AutoIt](https://chocolatey.org/packages/autoit.commandline), because AutoHotkey is more lightweight, FOSS, and is more actively developed than AutoIt.
```

##### Work on all locales

Script must work on every locale available for Windows, so be careful when using strings of text of window in the script that could be different in another language.

##### Avoid brittle scripts

> Do not create brittle scripts that work only when user doesn't interfer. All script elements should be as precise as possible - for instance, instead of using [Send](https://autohotkey.com/docs/commands/Send.htm) function which will work correctly only if the desired window is active, use [ControlSend](https://autohotkey.com/docs/commands/ControlSend.htm) which doesn't require window activation or use [BlockInput](https://autohotkey.com/docs/commands/BlockInput.htm) for short period of time.

Reviewer 3: ```suggestion
Do not create brittle scripts that work only when user doesn't interfere. All script elements should be as precise as possible - for instance, instead of using [Send](https://autohotkey.com/docs/commands/Send.htm) function which will work correctly only if the desired window is active, use [ControlSend](https://autohotkey.com/docs/commands/ControlSend.htm) which doesn't require window activation or use [BlockInput](https://autohotkey.com/docs/commands/BlockInput.htm) for short periods of time.
```

#### Source Files

##### Encoding

Always __use UTF-8 without BOM__ for the `*.nuspec` and __UTF-8 with BOM__ for the `*.ps1` files. See [character encodings](https://docs.chocolatey.org/en-us/create/create-packages#character-encoding).

##### Code style

> Do not commit code with obvious styling problems such as irregular indentation levels, very long lines, too many comments, too many empty lines, etc.

Reviewer 3: ```suggestion
Do not commit code with obvious styling problems such as irregular indentation levels, very long lines, too many comments, too many empty lines, etc. In general, please follow the [PoshCode PowerShell Practice and Style Guide](https://github.com/PoshCode/PowerShellPracticeAndStyle).
```

For manual packages, this is defined directly in the package metadata file. In general, this should always be in lowercase. However, existing packages should use the same casing as the original identifier.
This root directory name should always be in lowercase and follow the official [Chocolatey naming conventions](https://docs.chocolatey.org/en-us/create/create-packages#naming-your-package).

#### Dependency Versions

1. A minimum dependency version must be specified when adding a dependency. Without this, any version satisfies the dependency. That means it will only automatically upgrade to the latest version if someone upgrades the extension outside of their process or incidentally installs some package that uses an explicitly set newer version. Changes to a package that does not specify a minimum or exact version will not be accepted.
2. When creating a dependency for virtual packages, specify an exact version range for the dependent package (_.install_ or _.portable). This version should be the same as that of the virtual package. An exact version in the metadata file will look like `[1.0.0]`.

#### Conform To Guidelines

> Conform with the [official package creation guidelines](1754-Update-Guidelines-and-code-to-reflect-new-procedures) and take a look at the [quick start guide](https://docs.chocolatey.org/en-us/create/create-packages-quick-start) on how to create packages.

Author: Most of the places in this document uses markdown links. Just using raw URLs is only used for the bare minimal when it makes sense to do so.

One or more maintainers of the maintainers listed in the `owners` element should be equal to a similar name listed in our `CODE_OWNERS` file and will be considered the primary maintainer of the package.

#### Icons

A package must have an icon available if the packaged software has an icon. This icon must be named the same as the package and is placed in the [icons](https://github.com/chocolatey-community/chocolatey-packages/tree/master/icons) directory.
If the package name ends with either `.install` or `.portable`, the suffix may be ignored in the icon name.
When an icon is added to this folder with the correct name, it will **automatically** be set in the metadata file and the Readme file when our build server updates the package, if the metadata file contains an `<IconUrl>` element.

**IMPORTANT: If no icon is available, the comment `<!-- IconUrl: Skip check -->` should be added to the metadata file**

#### Chocolatey Compatibility

All packages are expected to be compatible with the Chocolatey CLI version, available 12 months earlier than the last change of the package.

> That means if you change the metadata or one of the package scripts **on January 12th, 2023**, it should be compatible with Chocolatey CLI 0.11.3 and later released versions.

Reviewer 4: Since we're now no longer supporting pre-1.0.0 versions, would it be worthwhile to update this date and version?

Author: As I have added a new Wiki Page for what we support, I'll update this section to link to that page.

If that is not possible, and the package needs a version of Chocolatey CLI released later, it will need to include a dependency for that Chocolatey CLI version.

### Chocolatey Scripts

> All PowerShell scripts are expected to be compatible with PowerShell v2 and later (not PowerShell Core). If this is not possible, a dependency on the package `PowerShell` is required, with its minimum version set to the earliest version that the scripts in the package are expected to work with.

Reviewer 4: ```suggestion
All PowerShell scripts are expected to be compatible with Windows PowerShell v2 through v5.1. If this is not possible, a dependency on the package `PowerShell` is required, with its minimum version set to the earliest version that the scripts in the package are expected to work with.
```

Is it worthwhile mentioning `Windows` PowerShell and being explicit about the versions instead of saying 2+ and then saying not Core (which technically 7.x isn't...)?

##### Code Style

Refrain from committing code with obvious styling problems such as irregular indentation levels, very long lines, too many comments, too many empty lines, and other styling issues. Please follow the [PoshCode PowerShell Practice and Style Guide](https://github.com/PoshCode/PowerShellPracticeAndStyle).

The project contains a [`.editorconfig`](https://github.com/chocolatey-community/chocolatey-packages/blob/master/.editorconfig)
 file that you can use with many editors via [EditorConfig](http://editorconfig.org/) plugins. Ensure you use an editor with this support and have the support enabled.

Keep the package source files clean and remove outdated code and unnecessary comments. Comment on non-obvious code so that others can easily understand what it does.

### Update Script

Update scripts are the PowerShell scripts responsible for getting information about the software's most current version, where it can be acquired from, and how to modify the source code to produce the package. It is typically called `update.ps1`.
These scripts should be located in the root of the package directory, right next to the metadata (`.nuspec`) file.

> The update script should be compatible with PowerShell v5 and may not need to be compatible with PowerShell Core.

Reviewer 4: ```suggestion
The update script should be compatible with Windows PowerShell v5 and may not need to be compatible with PowerShell Core.
```

I feel like it's cleaner to specify Windows PowerShell for anything before 6.0+, but I'm not sure the best way to refer to newer ones as 7.0 dropped the `Core` nomenclature, but most people still know it as Core (and generally it's understood what is meant).

Author: I think Core is still fine to reference it as. If you run `$PSVersionTable` it still says the edition being Core, if I recall correctly.

#### Use `UseBasicParsing` When Possible

If you use the PowerShell cmdlet `Invoke-WebRequest` in the update script, always make sure to also pass in `-UseBasicParsing` to this cmdlet. There may be cases where using this parameter is not possible. Add a comment before the call about why `-UseBasicParsing` can not be used in these cases.
Any script that does not use this argument requires the Internet Explorer engine to be available and the web browser to have launched at least once before.

#### Do Not Download Large Files

Unless a package installer/executable/archive needs some special handling (like reading the version from the file or something else not automatically handled by AU), do not download anything bigger than a few MB within `au_GetLatest` function. Usually, the files are downloaded during the updating process automatically or in the `au_BeforeUpdate` function.
Try finding alternatives to reading the version from the file when needed to prevent the need to download the file if no version elsewhere is available, like checking the headers of the response before doing a download.

## FAQ

### Why Do We Automatically Close Issues?

> The Chocolatey Packages repository is a shared repository that will be worked on by the entire Chocolatey Community. When an issue lives for a long enough time to become stale, it needs more interest in the Community for the issue to become fixed. As there needs to be more interest in the Community, it is unlikely the issue will be fixed anytime soon, and it will instead be automatically closed.

Reviewer 4: ```suggestion
The Chocolatey Packages repository is a shared repository that is worked on by the entire Chocolatey Community. When an issue lives for a long enough time to become stale, it needs more interest in the Community for the issue to become fixed. As there needs to be more interest in the Community, it is unlikely the issue will be fixed anytime soon, and it will instead be automatically closed.
```

Keep the package source files clean and remove obsolete or outdated code and unnecessary comments. Comment non-obvious code so that others can easily understand what it does.

### Update Script

> An update script are the PowerShell scripts responsible for getting the information about what the most current version of the software is, where it can be acquired from and is typically called `update.ps1`.

Reviewer 3: ```suggestion
Update scripts are the PowerShell scripts responsible for getting the information about what the most current version of the software is, where it can be acquired from, and how to modify the source code to produce the package. It is typically called `update.ps1`.
```

These scripts should be located in the root of the package directory, right next to the metadata (`.nuspec`) file.
The update script should be compatible with PowerShell v5, but may not need to be compatible with PowerShell Core.

#### Use `UseBasicParsing`

> If you use the PowerShell cmdlet in the update script, always make sure to also pass in `-UseBasicParsing` to this script. There may be cases where using this parameter is not possible, in these cases a comment right before the call should be specified about why `-UseBasicParsing` can not be used.

Reviewer 3: A discussion for another time:
Is there a reason we wouldn't _enforce_ UseBasicParsing by use of global variables and default parameters, and ensure that packages that do not require it _must specify it_ as `-UseBasicParsing:$false`? (Other than it being a bizarre pattern)

Author: I didn't know you could enforce it. Also, I think it can only be enforced partially (since each script is launched in its own job, I have sometimes seen that configured variables aren't carried over).

Any script that does not use this argument requires the Internet Explorer engine to be available, and the web browser to have launched at least once before.

#### Do not download large files

Unless a package installer/executable/archive needs some special handling (like the need to read the version from the file, or something else not automatically handled by AU), anything bigger than a few MB should never be downloaded within `au_GetLatest` function. Normally, the files are downloaded during the updating process automatically, or in the `au_BeforeUpdate` function.
Try finding alternatives to reading the version from the file when needed to prevent the need to download the file if no version elsewhere is available, like checking the headers of the response before doing a download.

## FAQ

### Why do we automatically close issues?

> The Chocolatey Packages repository is considered a shared repository that will be worked on by the entire Chocolatey Community. When an issue lives for long enough time to become stale, it is considered to not be enough interest in the community for the issue to become fixed. As such it is more appropriate to close the issue than continue to keep issues open that will not be worked on.

Reviewer 3: Though I understand our point, I wish this were not so.

### Why do not the repository maintainers fix the issues?

While there may be times that a repository maintainer will take steps to fix problems with a package, it is not the main reason for the repository maintainers here.

> The repository maintainers main tasks are to keep order in the repository, triage issues and making sure that pull requests becomes reviewed by the appropriate people and changes to packages is up to the quality expected for packages in the repository.

Reviewer 3: ```suggestion
The repository maintainers main tasks are to keep order in the repository, triage issues, make sure that pull requests are reviewed by the appropriate people, and changes to packages are up to the quality expected for packages in the repository.
```
